### PR TITLE
Enemies

### PR DESCRIPTION
**Pull request by ScottyXIII**

### CODE REVIEW

**Comment by ScottyXIII** [comment]
Enums folder starts with a capital. Needs to be changed along with any imports for it :( fucking windows.

**Comment by ScottyXIII** [comment]
Prob going to need to implement some sorta of Enemy factory to handle diff enemy types rather than setting a new spawnerConfig per type.

File: src/objects/entities/Hedgehog.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import * as Phaser from 'phaser';
+import Entity from '@/objects/Entities/Entity';
```
Comment (norgeous) [inline src/objects/entities/Hedgehog.ts:2]: change

`import Entity from '@/objects/Entities/Entity';`

to

`import Entity from '@/objects/entities/Entity';`

File: src/objects/entities/Entity.ts [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+      });
+    });
+    this.playAnimation("idle");
+     
+    // container
+    this.gameObject = this.scene.matter.add.gameObject(this);
+    this.scene.add.existing(this);
+    
+    // sensors
+    // @ts-ignor
```
Comment (norgeous) [inline src/objects/entities/Entity.ts:138]: typo

File: src/objects/entities/Entity.ts [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+    this.flipXSprite(this.facing === -1);
+
+    keepUpright(this.keepUprightStratergy, this.gameObject);
+
+    const { angularVelocity } = this.gameObject.body;
+    const speed = Math.hypot(this.gameObject.body.velocity.x, this.gameObject.body.velocity.y);
+    const motion = speed + Math.abs(angularVelocity);
+    const closeToStationary = motion <= 0.1;
+    // @ts-ignor
```
Comment (norgeous) [inline src/objects/entities/Entity.ts:181]: typo

File: src/objects/entities/Entity.ts [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    if (closeToStationary || this.constantMotion) {
+      const vectorTowardsPlayer = {
+        x: player.torso.x - this.x,
+        y: player.torso.y - this.y,
+      };
+      this.gameObject.setVelocity?.(
+        vectorTowardsPlayer.x < 0 ? -this.maxSpeedX : this.maxSpeedX,
+        vectorTowardsPlayer.y < 0 ? -this.maxSpeedY : this.maxSpeedY,
+      );
+    }
```
Comment (norgeous) [inline src/objects/entities/Entity.ts:200]: what does this do? move everything towards scene.player?

File: src/objects/entities/Entity.ts [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    });
+    this.playAnimation("idle");
+     
+    // container
+    this.gameObject = this.scene.matter.add.gameObject(this);
+    this.scene.add.existing(this);
+    
+    // sensors
+    // @ts-ignor
+    const { Bodies, Body } = Phaser.Physics.Matter.Matter;
```
Comment (norgeous) [inline src/objects/entities/Entity.ts:139]: ```const { bodies: Bodies, body: Body } = scene.matter;```
https://github.com/photonstorm/phaser3-docs/issues/31#issuecomment-1198668830

File: src/objects/Enums/Physics.ts [hunk 6]
```diff
@@ -0,0 +1,7 @@
+export default class keepUprightStratergies {
+  static NONE = 'NONE';
+
+  static INSTANT = 'INSTANT';
+
+  static SPRINGY = 'SPRINGY';
+}
```
Comment (norgeous) [inline src/objects/Enums/Physics.ts:7]: now that we have TS, this could be a real ENUM

File: src/objects/entities/Entity.ts [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+  bounce?: number,
+};
+
+type animationsConfigType = {
+  animationKey: string, 
+  start: number, 
+  end: number, 
+  fps: number,
+  repeat?: number | undefined
+};
```
Comment (norgeous) [inline src/objects/entities/Entity.ts:19]: i think we can pull these straight out of Phaser package

File: src/objects/entities/Entity.ts [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+type EntityConfigType = {
+  name: string,
+  spriteSheetKey: string,
+  animations?: animationsConfigType[],
+  physicsConfig?: physicsConfigType,
```
Comment (norgeous) [inline src/objects/entities/Entity.ts:25]: physicsConfig?: Phaser.Types.Physics.Matter.MatterBodyConfig;

File: src/objects/Enums/Physics.ts (hunk 6 above)
Comment (ScottyXIII) [reply src/objects/Enums/Physics.ts]: Changed to an enum but was throwing a bunch off errors for 'no-shadow' and 'un-used-vars', After some googling people suggested changing linting rules for no-shadow. There was also one for no-wars but I was still getting the error so I've supressed it in the enum file.
` "no-shadow": "off",
    "@typescript-eslint/no-shadow": "warn",`

File: src/objects/entities/Entity.ts (hunk 7 above)
Comment (ScottyXIII) [reply src/objects/entities/Entity.ts]: If we use Phasers built in Animation type it doesn't seem to have an endFrame param and requires we pass it an AnimationFrame object which is built from the function `anims.generateFrameNumbers()`. Might be easier to keep our custom type here or just use a custom type for the Frames. `end ` is only needed for `generateFrameNumbers` 

https://newdocs.phaser.io/docs/3.60.0/Phaser.Types.Animations

File: src/objects/entities/Entity.ts [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+import keepUpright from '@/helpers/keepUprightStratergy';
+import KeepUprightStratergies from '@/objects/Enums/KeepUprightStratergies';
+
+type AnimationsConfigType = {
+  animationKey: string;
+  start: number;
+  end: number;
+  fps: number;
+  repeat?: number | undefined;
+};
```
[the remark this answers is not in the document]
Comment (norgeous) [reply src/objects/entities/Entity.ts:11]: type AnimationsConfigType = Phaser.Types.Animations & {end:number}

[the remark this answers is not in the document]
Comment (norgeous) [reply src/objects/entities/Entity.ts:11]: or just pass in separate params as to not pollute existing types

**Review by norgeous — APPROVED** [review]

**Pull request merged**